File: src/main/java/com/meti/api/core/Option.java
package com.meti.api.core;

public interface Option<T> {
	<R> Option<R> map(F1<T, R> mapper);

	void ifPresent(Consumer<T> consumer);

	T orElse(T other);

	<R, E extends Exception> Option<R> mapExceptionally(EF1<T, R, E> mapper) throws E;

	<E extends Exception> T orElseThrow(Supplier<E> supplier) throws E;

	boolean isEmpty();

	boolean isPresent();

	<R> Option<R> flatMap(F1<T, Option<R>> mapper);

	Option<T> filter(F1<T, Boolean> predicate);

	T orElseGet(Supplier<T> supplier);
}

File: src/main/java/com/meti/api/core/Option.scala
package com.meti.api.core

import java.util.Optional
import java.util.function.Function
import java.util.function.Predicate
import java.util.function.Supplier

trait Option[T] {
  def toJava : Optional[T]

  def filter(predicate: Predicate[T]) : Option[T]

  def orElseSupply(supplier: Supplier[T]) : T

  def flatMap[R](function: Function[T, Option[R]]) : Option[R]

  def map[R](mapper: Function[T, R]) : Option[R]

  def orElse(other: T) : T
}